`NullPointerException` in `client.HttpResponseDecoder`

I got reported a NPE issue from [a Armeria Slack thread](https://line-armeria.slack.com/archives/C1NGPBUH2/p1627494312253700?thread_ts=1626604451.208400&cid=C1NGPBUH2).
```
Caused by:java.lang.NullPointerException: Cannot invoke "com.linecorp.armeria.client.HttpResponseDecoder$HttpResponseWrapper.close(java.lang.Throwable)" because "res" is null(Cannot invoke "com.linecorp.armeria.client.HttpResponseDecoder$HttpResponseWrapper.close(java.lang.Throwable)" because "res" is null)
  at com.linecorp.armeria.client.HttpResponseDecoder.failUnfinishedResponses(HttpResponseDecoder.java:131)
  at com.linecorp.armeria.client.HttpSessionHandler.channelInactive(HttpSessionHandler.java:415)
```

`HttpResponseDecoder` MUST be executed in an event loop. The NPE might be caused by recursive calls of `failUnfinishedResponses()`.
Even though we use `Iterator`, a null reference could be returned.
```java
private final IntObjectMap<String> intObjectHashMap = new IntObjectHashMap<>();
private final Map<Integer, String> hashMap = new HashMap<>();

@Test
void intObjectHashMap() {
    intObjectHashMap.put(1, "A");
    intObjectHashMap.put(2, "B");
    intObjectHashMap.put(3, "C");
    intObjectHashMap.put(4, "D");
    intObjectHashMap.put(5, "E");
    // java.lang.AssertionError: null is not a legitimate internal value. Concurrent Modification?
    removeAll(intObjectHashMap);
}

@Test
void jdkHashMap() {
    hashMap.put(1, "A");
    hashMap.put(2, "B");
    hashMap.put(3, "C");
    hashMap.put(4, "D");
    hashMap.put(5, "E");
    // java.util.ConcurrentModificationException
    removeAll(hashMap);
}

void removeAll(Map<Integer, String> strings) {
    for (final Iterator<String> iterator = strings.values().iterator();
         iterator.hasNext();) {
        final String res = iterator.next();
        System.out.println("removed: " + res);
        // To avoid calling removeResponse by res.close(cause), remove before closing.
        iterator.remove();
        removeAll(strings);
    }
```

--- a/core/src/main/java/com/linecorp/armeria/client/HttpResponseDecoder.java
+++ b/core/src/main/java/com/linecorp/armeria/client/HttpResponseDecoder.java
@@ -58,7 +58,7 @@ abstract class HttpResponseDecoder {
 
     private int unfinishedResponses;
     private boolean disconnectWhenFinished;
-    private int wip;
+    private boolean closing;
 
     HttpResponseDecoder(Channel channel, InboundTrafficController inboundTrafficController) {
         this.channel = channel;
@@ -101,6 +101,11 @@ abstract class HttpResponseDecoder {
 
     @Nullable
     final HttpResponseWrapper removeResponse(int id) {
+        if (closing) {
+            // `unfinishedResponses` will be removed by `failUnfinishedResponses()`
+            return null;
+        }
+
         final HttpResponseWrapper removed = responses.remove(id);
         if (removed != null) {
             unfinishedResponses--;
@@ -123,21 +128,19 @@ abstract class HttpResponseDecoder {
     }
 
     final void failUnfinishedResponses(Throwable cause) {
-        if (wip++ > 0) {
-            // re-entrance
+        if (closing) {
             return;
         }
+        closing = true;
 
-        do {
-            for (final Iterator<HttpResponseWrapper> iterator = responses.values().iterator();
-                 iterator.hasNext();) {
-                final HttpResponseWrapper res = iterator.next();
-                // To avoid calling removeResponse by res.close(cause), remove before closing.
-                iterator.remove();
-                unfinishedResponses--;
-                res.close(cause);
-            }
-        } while (--wip != 0);
+        for (final Iterator<HttpResponseWrapper> iterator = responses.values().iterator();
+             iterator.hasNext();) {
+            final HttpResponseWrapper res = iterator.next();
+            // To avoid calling removeResponse by res.close(cause), remove before closing.
+            iterator.remove();
+            unfinishedResponses--;
+            res.close(cause);
+        }
     }
 
     abstract KeepAliveHandler keepAliveHandler();

--- a/core/src/main/java/com/linecorp/armeria/client/HttpResponseDecoder.java
+++ b/core/src/main/java/com/linecorp/armeria/client/HttpResponseDecoder.java
@@ -58,6 +58,7 @@ abstract class HttpResponseDecoder {
 
     private int unfinishedResponses;
     private boolean disconnectWhenFinished;
+    private int wip;
 
     HttpResponseDecoder(Channel channel, InboundTrafficController inboundTrafficController) {
         this.channel = channel;
@@ -122,14 +123,21 @@ abstract class HttpResponseDecoder {
     }
 
     final void failUnfinishedResponses(Throwable cause) {
-        for (final Iterator<HttpResponseWrapper> iterator = responses.values().iterator();
-             iterator.hasNext();) {
-            final HttpResponseWrapper res = iterator.next();
-            // To avoid calling removeResponse by res.close(cause), remove before closing.
-            iterator.remove();
-            unfinishedResponses--;
-            res.close(cause);
+        if (wip++ > 0) {
+            // re-entrance
+            return;
         }
+
+        do {
+            for (final Iterator<HttpResponseWrapper> iterator = responses.values().iterator();
+                 iterator.hasNext();) {
+                final HttpResponseWrapper res = iterator.next();
+                // To avoid calling removeResponse by res.close(cause), remove before closing.
+                iterator.remove();
+                unfinishedResponses--;
+                res.close(cause);
+            }
+        } while (--wip != 0);
     }
 
     abstract KeepAliveHandler keepAliveHandler();

